crypto/fipsmodule/evp/p_hmac.c

```diff
@@ -0,0 +1,230 @@
+
+static int hmac_init(EVP_PKEY_CTX *ctx) {
+  HMAC_PKEY_CTX *hctx;
+  hctx = OPENSSL_malloc(sizeof(HMAC_PKEY_CTX));
+  if (hctx == NULL) {
+    return 0;
+  }
+  OPENSSL_memset(hctx, 0, sizeof(HMAC_PKEY_CTX));
+  HMAC_CTX_init(&hctx->ctx);
+  ctx->data = hctx;
+  return 1;
+}
+
+static int hmac_copy(EVP_PKEY_CTX *dst, EVP_PKEY_CTX *src) {
+  HMAC_PKEY_CTX *sctx, *dctx;
+  if (!hmac_init(dst)) {
+    return 0;
+  }
+  sctx = src->data;
+  dctx = dst->data;
+  dctx->md = sctx->md;
+  HMAC_CTX_init(&dctx->ctx);
+  if (!HMAC_CTX_copy_ex(&dctx->ctx, &sctx->ctx)) {
+    return 0;
+  }
+
+  if (sctx->key_len != 0) {
+    dctx->key = OPENSSL_memdup(sctx->key, sctx->key_len);
+    if (dctx->key == NULL) {
+      return 0;
+    }
+    dctx->key_len = sctx->key_len;
+  }
+  return 1;
+}
+
+static void hmac_cleanup(EVP_PKEY_CTX *ctx) {
+  HMAC_PKEY_CTX *hctx = ctx->data;
+  if (hctx == NULL) {
+    return;
+  }
+  HMAC_CTX_cleanse(&hctx->ctx);
+  if (hctx->key != NULL) {
+    OPENSSL_free(hctx->key);
+  }
+  OPENSSL_free(hctx);
+}
+
+static int hmac_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
+  HMAC_PKEY_CTX *hctx = ctx->data;
+  CBB *key = OPENSSL_malloc(sizeof(CBB));
+  if (key == NULL) {
+    return 0;
+  }
+
+  if (!CBB_init(key, 0) || !CBB_add_bytes(key, hctx->key, hctx->key_len)) {
+    CBB_cleanup(key);
+    return 0;
+  }
+
+  // |EVP_PKEY_new_mac_key| allocates a temporary |EVP_PKEY_CTX| and frees it
+  // along with |ctx->data| (which contains the original key data).
+  // We allocate and copy the key data over to a new |CBB| and assign it to
+  // |pkey|, so that it is available in later operations.
+  return EVP_PKEY_assign(pkey, EVP_PKEY_HMAC, key);
+}
+
+static void hmac_update(EVP_MD_CTX *ctx, const void *data, size_t count) {
+  HMAC_PKEY_CTX *hctx = ctx->pctx->data;
+  HMAC_Update(&hctx->ctx, data, count);
+}
+
+static int hmac_init_set_up(EVP_PKEY_CTX *ctx, EVP_MD_CTX *mctx) {
+  // |mctx| gets repurposed as a hook to call |HMAC_Update|. |mctx->update| is
+  // normally copied from |mctx->digest->update|, but |EVP_PKEY_HMAC| has its
+  // own definition. We suppress the automatic setting of |mctx->update| and the
+  // rest of its initialization here.
+  mctx->flags |= EVP_MD_CTX_FLAG_NO_INIT_FOR_HMAC;
+  mctx->update = hmac_update;
+  return 1;
+}
+
+static int hmac_final(EVP_PKEY_CTX *ctx, uint8_t *sig, size_t *siglen,
+                      EVP_MD_CTX *mctx) {
+  unsigned int hlen;
+  HMAC_PKEY_CTX *hctx = ctx->data;
+  size_t md_size = EVP_MD_CTX_size(mctx);
+
+  if (sig == NULL) {
+    *siglen = md_size;
+    return 1;
+  } else if (*siglen < md_size) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  if (!HMAC_Final(&hctx->ctx, sig, &hlen)) {
+    return 0;
+  }
+  *siglen = (size_t)hlen;
+  return 1;
+}
+
+static int hmac_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2) {
+  HMAC_PKEY_CTX *hctx = ctx->data;
+
+  switch (type) {
+    case EVP_PKEY_CTRL_HMAC_SET_MAC_KEY: {
+      const CBS *key = p2;
```

Comment:
`HMAC_PKEY_CTX` already have a key length. I don't think there is a reason to wrap this in a CBB buffer. We are just writing and reading a clean byte key.